source/neuropod/bindings/java/src/main/java/com/uber/neuropod/NeuropodTensor.java

```diff
@@ -106,110 +107,187 @@ public FloatBuffer toFloatBuffer() {
+    }
+
+    /**
+     * Gets an int element at the given index.
+     * <p>
+     * Can only be used when the tensor is INT32_TENSOR.
+     * For example, to get an int element from a 3rd-order tensor,
+     * the user should call getInt(x, y, z)
+     *
+     * @param index the index array
+     * @return the int element
+     */
+    public int getInt(long... index) {
+        checkType(TensorType.INT32_TENSOR);
+        long pos = toPos(index);
+        return buffer.asIntBuffer().get((int) pos);
+    }
+
+    /**
+     * Gets a long element at the given index.
+     * <p>
+     * Can only be used when the tensor is INT64_TENSOR.
+     * For example, to get an long element from a 3rd-order tensor,
+     * the user should call getLong(x, y, z)
+     *
+     * @param index the index array
+     * @return the long element
+     */
+    public long getLong(long... index) {
+        checkType(TensorType.INT64_TENSOR);
+        long pos = toPos(index);
+        return buffer.asLongBuffer().get((int) pos);
+    }
+
+    /**
+     * Gets a double element at the given index.
+     * <p>
+     * Can only be used when the tensor is DOUBLE_TENSOR.
+     * For example, to get a double element from a 3rd-order tensor,
+     * the user should call getDouble(x, y, z)
+     *
+     * @param index the index array
+     * @return the double element
+     */
+    public double getDouble(long... index) {
+        checkType(TensorType.DOUBLE_TENSOR);
+        long pos = toPos(index);
+        return buffer.asDoubleBuffer().get((int) pos);
+    }
+
+    /**
+     * Gets a float element at the given index.
+     * <p>
+     * Can only be used when the tensor is FLOAT_TENSOR.
+     * For example, to get a float element from a 3rd-order tensor,
+     * the user should call getFloat(x, y, z)
+     *
+     * @param index the index array
+     * @return the float element
+     */
+    public float getFloat(long... index) {
+        checkType(TensorType.FLOAT_TENSOR);
+        long pos = toPos(index);
+        return buffer.asFloatBuffer().get((int) pos);
+    }
+
+    /**
+     * Gets a string element at the given index.
+     * <p>
+     * Can only be used when the tensor is STRING_TENSOR.
+     * For example, to get a string element from a 3rd-order tensor,
+     * the user should call getString(x, y, z)
+     *
+     * @param index the index array
+     * @return the string element
+     */
+    public String getString(long... index) {
+        checkType(TensorType.STRING_TENSOR);
+        long pos = toPos(index);
+        return nativeGetString(pos, super.getNativeHandle());
+    }
 
     private void checkType(TensorType type) {
         if (getTensorType() != type) {
-            throw new NeuropodJNIException("TensorType mismatch! Expected " +
+            throw new NeuropodJNIException("tensorType mismatch, expected " +
                     type.name() + ", found " + getTensorType().name());
 
         }
     }
 
+    private long toPos(long[] index) {
```

Comment:
Please add test coverage for this (using tensors with somewhat complex dimensions)

Reply:
Yes, I will add necessary code to this PR and add test coverage that covers toPos.

Reply:
Before I was going to create a last PR with Tensor Allocator that had all functionality and so Unit tests. I decided to add it and Unit Tests to this PR.

PR will become slightly bigger (though I simplified code with get_data_at/set_data_at and removed some code).  But this way it will be more consistent and we will see what is covered and what missed better in this PR.